---
title: "R Notebook"
output: html_notebook
---


```{r}
#install.packages("tidyverse")
#install.packages("purrr")
#install.packages("dplyr")
#install.packages("plyr")

library(tidyverse)
library(purrr)
library(dplyr)
library(plyr)
```

# Gene Level - TPM Data
```{r}
FileNames = c("944478", "944387", "944388", "944389", "944390", "944391", "944392", "944394", "944396", "944397", "944398", "944400", "944401", "944402", "944403", "944404", "944405", "944406", "944407", "944408", "944409", "944410", "944411", "944412", "944414", "944415", "944416", "944417", "944418", "944419", "944420", "944421", "944422", "944423", "944424", "944425", "944426", "944427", "944429", "944430", "944431", "944432", "944434", "944435", "944436", "944437", "944438", "944439", "944440", "944441", "944442", "944443", "944444", "944445", "944446", "944447", "944448", "944449", "944450", "944452", "944453", "944454", "944455", "944456", "944457", "944458", "944459", "944461", "944462", "944463", "944464", "944466", "944467", "944468", "944470", "944471", "944472", "944473", "944474", "944475", "944476", "944477")

SampleNames = read.csv("~/Dropbox/Columbia/Mouse Punch RNAseq/Input Data/SampleNames.csv", head=T)

##TPM DATA
Temp = read.table(paste("~/Dropbox/Columbia/Mouse Punch RNAseq/Input Data/Kallisto Files/kallisto.sample.", "944478", "/abundance.genes.txt", sep=""), head = T)
Compiled_TPM = Temp
Compiled_TPM[2] = NULL
colnames(Compiled_TPM) = "Gene"

for(fn in FileNames){
Data = read.table(paste("~/Dropbox/Columbia/Mouse Punch RNAseq/Input Data/Kallisto Files/kallisto.sample.", fn, "/abundance.genes.txt", sep=""), head = T)
GetNumber = paste("LC", gsub("kallisto.sample.", "", fn), sep="")
FindFileName = subset(SampleNames, SampleNames$target_id %in% GetNumber)
colnames(Data) = c("Gene", FindFileName[1,2])
Compiled_TPM = merge(Compiled_TPM, Data, by="Gene")
}

Compiled_TPM_OutliersRM = subset(Compiled_TPM, select = -c(`PVH (Fresh 5wk)_1`, `PVH (Fresh 5wk)_2`, `PVH (Fresh 5wk)_3`, `PVH (Frozen 5wk)_1`, `PVH (Frozen 5wk)_2`, `PVH (Frozen 5wk)_3`, DVC_1, ENT_1, MB_2, PBN_3, VMH_1, PVH_2))
colnames(Compiled_TPM_OutliersRM) = gsub("DVC_4", "DVC_1", gsub("ENT_4", "ENT_1", gsub("MB_3", "MB_2", gsub("PBN_4", "PBN_3", gsub("PVH_3", "PVH_2", gsub("VMH_3", "VMH_1", colnames(Compiled_TPM_OutliersRM)))))))

colnames(Compiled_TPM_OutliersRM) = gsub("LH_", "LHA_", gsub("PFC_", "FRP_", gsub("ACC_", "ACA_", gsub("Nac_", "ACB_", gsub("ARC_", "ARH_", gsub("PBN_", "PB_",  gsub("AV_", "CENT2_", gsub("PV_", "CUL4,5_", gsub("BV_", "UVU_", gsub("CN_", "CBN_", gsub("HYP_", "THy/PHy_", gsub("FB_", "F_", gsub("MB_", "M_", gsub("HB_", "H_", gsub("FB_", "F_", gsub("B6A.mESC_", "mESC B6-1_", gsub("B6E.mESC_", "mESC B6-2_", colnames(Compiled_TPM_OutliersRM))))))))))))))))))

Compiled_TPM_OutliersRM = Compiled_TPM_OutliersRM %>% tibble %>% select("Gene", sort(colnames(.)))
write.csv(Compiled_TPM_OutliersRM, "~/Dropbox/Columbia/Mouse Punch RNAseq/DataForGEO/DoegeLab_BrainPunch_TPMData_GeneLevel.csv")
```

# Transcript_Level - COUNTS AND TPM
```{r}
Temp = read.table(paste("~/Dropbox/Columbia/Mouse Punch RNAseq/Input Data/Kallisto Files/kallisto.sample.", "944478", "/abundance.tsv", sep=""), head = T)
Compiled_TranscriptCounts = Temp
Compiled_TranscriptCounts[2:length(colnames(Compiled_TranscriptCounts))] = NULL
colnames(Compiled_TranscriptCounts) = "TranscriptID"


Compiled_TranscriptTPM = Temp
Compiled_TranscriptTPM[2:length(colnames(Compiled_TranscriptTPM))] = NULL
colnames(Compiled_TranscriptTPM) = "TranscriptID"

for(fn in FileNames){
Data = read.table(paste("~/Dropbox/Columbia/Mouse Punch RNAseq/Input Data/Kallisto Files/kallisto.sample.", fn, "/abundance.tsv", sep=""), head = T)
GetNumber = paste("LC", gsub("kallisto.sample.", "", fn), sep="")
FindFileName = subset(SampleNames, SampleNames$target_id %in% GetNumber)
Data1 = Data %>% select(target_id, est_counts)
colnames(Data1) = c("TranscriptID", FindFileName[1,2])
Compiled_TranscriptCounts = merge(Compiled_TranscriptCounts, Data1, by="TranscriptID")

Data2 = Data %>% select(target_id, tpm)
colnames(Data2) = c("TranscriptID", FindFileName[1,2])
Compiled_TranscriptTPM = merge(Compiled_TranscriptTPM, Data2, by="TranscriptID")
}

Compiled_TranscriptTPM_OutliersRM = subset(Compiled_TranscriptTPM, select = -c(`PVH (Fresh 5wk)_1`, `PVH (Fresh 5wk)_2`, `PVH (Fresh 5wk)_3`, `PVH (Frozen 5wk)_1`, `PVH (Frozen 5wk)_2`, `PVH (Frozen 5wk)_3`, DVC_1, ENT_1, MB_2, PBN_3, VMH_1, PVH_2))
colnames(Compiled_TranscriptTPM_OutliersRM) = gsub("DVC_4", "DVC_1", gsub("ENT_4", "ENT_1", gsub("MB_3", "MB_2", gsub("PBN_4", "PBN_3", gsub("PVH_3", "PVH_2", gsub("VMH_3", "VMH_1", colnames(Compiled_TranscriptTPM_OutliersRM)))))))

colnames(Compiled_TranscriptTPM_OutliersRM) = gsub("LH_", "LHA_", gsub("PFC_", "FRP_", gsub("ACC_", "ACA_", gsub("Nac_", "ACB_", gsub("ARC_", "ARH_", gsub("PBN_", "PB_",  gsub("AV_", "CENT2_", gsub("PV_", "CUL4,5_", gsub("BV_", "UVU_", gsub("CN_", "CBN_", gsub("HYP_", "THy/PHy_", gsub("FB_", "F_", gsub("MB_", "M_", gsub("HB_", "H_", gsub("FB_", "F_", gsub("B6A.mESC_", "mESC B6-1_", gsub("B6E.mESC_", "mESC B6-2_", colnames(Compiled_TranscriptTPM_OutliersRM))))))))))))))))))

Compiled_TranscriptTPM_OutliersRM = Compiled_TranscriptTPM_OutliersRM %>% tibble %>% select("TranscriptID", sort(colnames(.)))
write.csv(Compiled_TranscriptTPM_OutliersRM, "~/Dropbox/Columbia/Mouse Punch RNAseq/DataForGEO/DoegeLab_BrainPunch_TPMData_TranscriptLevel.csv")


Compiled_TranscriptCounts_OutliersRM = subset(Compiled_TranscriptCounts, select = -c(`PVH (Fresh 5wk)_1`, `PVH (Fresh 5wk)_2`, `PVH (Fresh 5wk)_3`, `PVH (Frozen 5wk)_1`, `PVH (Frozen 5wk)_2`, `PVH (Frozen 5wk)_3`, DVC_1, ENT_1, MB_2, PBN_3, VMH_1, PVH_2))
colnames(Compiled_TranscriptCounts_OutliersRM) = gsub("DVC_4", "DVC_1", gsub("ENT_4", "ENT_1", gsub("MB_3", "MB_2", gsub("PBN_4", "PBN_3", gsub("PVH_3", "PVH_2", gsub("VMH_3", "VMH_1", colnames(Compiled_TranscriptCounts_OutliersRM)))))))

colnames(Compiled_TranscriptCounts_OutliersRM) = gsub("LH_", "LHA_", gsub("PFC_", "FRP_", gsub("ACC_", "ACA_", gsub("Nac_", "ACB_", gsub("ARC_", "ARH_", gsub("PBN_", "PB_",  gsub("AV_", "CENT2_", gsub("PV_", "CUL4,5_", gsub("BV_", "UVU_", gsub("CN_", "CBN_", gsub("HYP_", "THy/PHy_", gsub("FB_", "F_", gsub("MB_", "M_", gsub("HB_", "H_", gsub("FB_", "F_", gsub("B6A.mESC_", "mESC B6-1_", gsub("B6E.mESC_", "mESC B6-2_", colnames(Compiled_TranscriptCounts_OutliersRM))))))))))))))))))

Compiled_TranscriptCounts_OutliersRM = Compiled_TranscriptCounts_OutliersRM %>% tibble %>% select("TranscriptID", sort(colnames(.)))
write.csv(Compiled_TranscriptCounts_OutliersRM, "~/Dropbox/Columbia/Mouse Punch RNAseq/DataForGEO/DoegeLab_BrainPunch_CountsData_TranscriptLevel.csv")
```

#Convert Transcript Counts dataframe to Gene names
```{r}
BiocManager::install("biomaRt")
library(biomaRt)
mart <- biomaRt::useMart(biomart = "ENSEMBL_MART_ENSEMBL",
  dataset = "mmusculus_gene_ensembl",
  host = "https://Oct2018.archive.ensembl.org") #Oct2016
 #  host = "https://ensembl.org")
ttg <- biomaRt::getBM(
  attributes = c("ensembl_transcript_id", "external_gene_name"),
  mart = mart)

TranscriptDF = Compiled_TranscriptCounts_OutliersRM
TranscriptDF$TranscriptID = gsub("\\..*", "", TranscriptDF$TranscriptID)

TranscriptDFmerge = merge(ttg, TranscriptDF, by.x = "ensembl_transcript_id", by.y = "TranscriptID")
TranscriptDFmerge$ensembl_transcript_id = NULL

Counts_Gene = TranscriptDFmerge %>% group_by(external_gene_name) %>% summarize_all(sum)
```
